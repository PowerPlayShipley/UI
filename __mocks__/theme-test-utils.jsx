// TODO: Fix this??

import * as React from 'react'
import userEvent from '@testing-library/user-event'
import { render as rtlRender } from '@testing-library/react'

import { ThemeProvider } from "styled-components";

import theme from '../src/themes'
import GlobalStyle from "../src/components/GlobalStyle";

function renderWithTheme(ui, { ...options } = {}) {
  const Wrapper = ({ children }) => (
    <ThemeProvider theme={theme}>
      <GlobalStyle />
      {children}
      <div id='modal-root' />
    </ThemeProvider>
  )

  return rtlRender(ui, { wrapper: Wrapper, ...options })
}

// THE BELLOW COME FROM `https://github.com/styled-components/styled-components/blob/0aa3170c255a49cd41c3fbeb2b8051b5d132f229/src/test/utils.js`

const stripComments = (str) => str.replace(/\/\*.*?\*\/\n?/g, '');

const stripWhitespace = (str) => str
  .trim()
  .replace(/([;\{\}])/g, '$1  ')
  .replace(/\s+/g, ' ');

const getCSS = (scope) =>
  Array.from(scope.querySelectorAll('style'))
    .map(tag => tag.innerHTML)
    .join('\n')
    .replace(/ {/g, '{')
    .replace(/:\s+/g, ':')
    .replace(/:\s+;/g, ':;');

function expectCSSMatches ( _expectation, opts = { ignoreWhitespace: true }) {
  // NOTE: This should normalise both CSS strings to make irrelevant mismatches less likely
  const expectation = _expectation
    .replace(/ {/g, '{')
    .replace(/:\s+/g, ':')
    .replace(/:\s+;/g, ':;');

  const css = getCSS(document);

  if (opts.ignoreWhitespace) {
    const stripped = stripWhitespace(stripComments(css));
    expect(stripped).toEqual(stripWhitespace(expectation));
    return stripped;
  } else {
    expect(css).toEqual(expectation);
    return css;
  }
};

export * from '@testing-library/react';

export { renderWithTheme, expectCSSMatches };
