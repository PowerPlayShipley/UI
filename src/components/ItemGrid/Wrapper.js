import styled from "styled-components";

const Wrapper = styled.div`
  display: flex;
  flex: 1;
  
  // Very crude but will work
  div[disabled],  div.disabled, div[aria-disabled=true]  {
    pointer-events: none;
    
    color: ${({ theme }) => theme.main.colors.text.disabled };
    opacity: 65%;
  }
`

Wrapper.displayName = 'Grid'

export default Wrapper
