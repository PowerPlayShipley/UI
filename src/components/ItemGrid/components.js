import styled from "styled-components";

export const Icon = styled.div`
  display: flex;
  align-items: center;
  
  height: 24px;
  width: 24px;
  
  justify-content: center;
`

export const Item = styled.div`
  display: flex;
  align-items: center;
  
  border-radius: 3px;
  padding: 4px;
  
  color: #222f3e;
  
  cursor: pointer;
  user-select: none;
  
  &:hover {
    background-color: #dee0e2;
  }
`

export const Grid = styled.div`
  display: flex;
  flex-wrap: wrap;
  
  max-height: 200px;
  
  overflow-x: hidden;
  overflow-y: auto;
  
  padding: 0;
`
