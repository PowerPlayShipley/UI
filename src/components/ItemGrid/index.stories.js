import ItemGrid from "./index";

export default {
  title: 'ItemGrid',
  component: ItemGrid
}

const Template = ({ items, length, ...rest }) => {
  const arr = Array.from({ length }, () => items[getRandomInt(items.length)])
  return (<ItemGrid items={arr} {...rest} />);
}

export const Default = Template.bind({});
Default.args = {
  length: 100,
  disabled: false,
  items: [ "😴", "😄", "😃", "⛔", "🎠", "🚓", "🚇"]
}

export const Tooltip = Template.bind({})
Tooltip.args = {
  length: 50,
  disabled: false,
  items: [{
    value: "😀",
    tooltip: 'Grinning face'
  }, {
    value: '🥰',
    tooltip: 'Smiling face with hearts'
  }, {
    value: '👁',
    tooltip: 'Eye'
  }, {
    value: '🌗',
    tooltip: 'Last quarter moon'
  }, {
    value: '🌎',
    tooltip: 'Globe showing Americas'
  }]
}

export const DisabledTooltip = Template.bind({})
DisabledTooltip.args = {
  length: 50,
  disabled: true,
  items: [{
    value: "😀",
    tooltip: 'Grinning face'
  }, {
    value: '🥰',
    tooltip: 'Smiling face with hearts'
  }, {
    value: '👁',
    tooltip: 'Eye'
  }, {
    value: '🌗',
    tooltip: 'Last quarter moon'
  }, {
    value: '🌎',
    tooltip: 'Globe showing Americas'
  }]
}

export const CustomisedTooltip = Template.bind({})
CustomisedTooltip.args = {
  length: 50,
  disabled: false,
  items: [{
    value: "😀",
    tooltip: 'Grinning face'
  }, {
    value: '🥰',
    tooltip: 'Smiling face with hearts'
  }, 'Long Word', {
    value: '🌗',
    tooltip: 'Last quarter moon'
  }, {
    value: '🌎',
    tooltip: 'Globe showing Americas',
    disabled: true
  }]
}

function getRandomInt(max) {
  return Math.floor(Math.random() * max);
}
