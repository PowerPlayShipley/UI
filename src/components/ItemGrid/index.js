/**
 * ItemGrid:
 *
 * Create an item Grid, this can be created passing just values along, or with a child function.
 * Using the function allows for customisation, i.e adding prefix/suffix if the grid requires it
 * */
import React, { useCallback } from 'react'
import PropTypes from 'prop-types'
import clsx from "clsx";

import Wrapper from "./Wrapper";
import { Icon, Item, Group } from './components'

import tooltip from "../../utils/tooltip";

// Get an item from element if an object or return the element itself
const get = (key, element) => (typeof element === 'object' && element[key]) ? element[key] : element
// Get the className
const getClassName = (element) => (typeof element === 'object' && element.className) ? element.className : null

// if an element is disabled independently it should not affect the rest
const disabled = (element, global) => global === true
  ? true : (typeof element === 'object' && element.disabled)
    ? element.disabled : false

const ItemGrid = ({ className, items, disabled: isDisabled = false, onClick, children }) => {
  // Only pass up if the click exists
  const handleClick = useCallback((e, item, idx) => {
    e.preventDefault()
    onClick && onClick(item, idx)
  }, [onClick])

  // Helper to allow children to add extra items to the grid outside the passed items
  const createItem = (item, idx, arr) => {
    return item && (
      <Item className={clsx(getClassName(item))} data-item={get('name', item)} {...tooltip(item)} key={idx}
            aria-disabled={disabled(item, isDisabled)} onClick={(e) =>  typeof arr === 'function' ? arr(e, item, idx) : handleClick(e, item, idx)}>
        <Icon>{get('value', item)}</Icon>
      </Item>
    )
  }

  return (
    <Wrapper className={clsx(className)}>
      <Group>
        {children ? children({
          items: () => items.map(createItem), createItem
        }) : items.map(createItem)}
      </Group>
    </Wrapper>
  )
}

const ValueProps = PropTypes.oneOfType([
  PropTypes.string,
  PropTypes.shape({
    value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.element
    ]).isRequired,
    tooltip: PropTypes.string
  })
])

ItemGrid.propTypes = {
  className: PropTypes.string,
  children: PropTypes.func,
  items: PropTypes.arrayOf(ValueProps).isRequired,
  disabled: PropTypes.bool,
  onClick: PropTypes.func
}

export default ItemGrid
