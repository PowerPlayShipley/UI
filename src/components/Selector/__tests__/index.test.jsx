/**
 * @jest-environment jsdom
 */

import React from 'react'
import { fireEvent, renderWithTheme } from '../../../../__mocks__/theme-test-utils'

import { Selector } from "../index";
import ItemGrid from "../../ItemGrid";

describe('Display Button', () => {
  it('should render cell', () => {
    const { container } = renderWithTheme(<Selector title={'Hello'} items={[]}/>)
    expect(container).toHaveTextContent('Hello')
  })

  it('should create valid snapshot', function () {
    const { container } = renderWithTheme(<Selector title={'Hello'} items={[]}/>)
    expect(container.firstChild).toMatchSnapshot()
  });

  it('should load correct amount of items', function () {
    const items = ['1', '2', '3', '4']
    const { container, getByText } = renderWithTheme(<Selector title={'Hello'} items={items} />)

    for (const item of items) {
      const child = getByText(item)
      expect(container).toContainElement(child)
    }
  });

  it('should fire an event on item clicked', function () {
    const handleClick = jest.fn()

    const { getByText } = renderWithTheme(<Selector items={['1']} title={'Hello'} onItemClicked={handleClick} />)

    fireEvent.click(getByText('1').parentElement)
    expect(handleClick).toHaveBeenCalledTimes(1)
  });

  it('should fire an event on close clicked', function () {
    const handleClick = jest.fn()

    const { getByText } = renderWithTheme(<Selector items={['1']} title={'Hello'} onCloseClicked={handleClick} />)

    fireEvent.click(getByText('Close'))
    expect(handleClick).toHaveBeenCalledTimes(1)
  });
})
