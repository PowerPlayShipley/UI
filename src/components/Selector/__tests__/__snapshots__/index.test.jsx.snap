// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display Button should create valid snapshot 1`] = `
.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  min-width: 0;
  text-align: left;
  text-transform: none;
  color: #24292f;
}

.c4 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  max-height: 650px;
  overflow: auto;
  padding: 16px 16px;
}

.c1 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  border-bottom: none;
  color: #24292f;
  padding: 16px 16px 0 16px;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  border-top: 1px rgba(208,215,222,0.5) solid;
  padding: 8px 16px;
}

.c7 button {
  margin-left: 8px;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: hidden;
  background-color: #ffffff;
  border: 1px rgba(208,215,222,0.5) solid;
  border-radius: 6px;
  box-shadow: 0 1px 0 rgba(27,31,36,0.04);
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c5 div[disabled],
.c5 div.disabled,
.c5 div[aria-disabled=true] {
  pointer-events: none;
  color: #8c959f;
  opacity: 65%;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  max-height: 200px;
  overflow-x: hidden;
  overflow-y: auto;
  padding: 0;
}

.c9 {
  position: relative;
  display: inline-block;
  padding: 5px 16px;
  font-size: 14px;
  font-weight: 500;
  line-height: 20px;
  white-space: nowrap;
  vertical-align: middle;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border: 1px solid;
  border-radius: 6px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c9:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c9:disabled,
.c9.disabled,
.c9[aria-disabled=true] {
  cursor: default;
}

.c9 .dropdown-caret {
  margin-left: 4px;
  opacity: 0.8;
}

.c9 {
  color: #24292f;
  background-color: #f6f8fa;
  border: 1px rgba(208,215,222,0.5) solid;
  box-shadow: 0 1px 0 rgba(27,31,36,0.04),inset 0 1px 0 rgba(255,255,255,0.25);
  -webkit-transition: 0.2s cubic-bezier(0.3,0,0.5,1);
  transition: 0.2s cubic-bezier(0.3,0,0.5,1);
  -webkit-transition-property: color,background-color,border-color;
  transition-property: color,background-color,border-color;
}

.c9:hover,
.c9.hover,
[open] > .c9 {
  background-color: #eaeef2;
  -webkit-transition-duration: 0.1s;
  transition-duration: 0.1s;
}

.c9:active {
  background-color: #e1e6ec;
  -webkit-transition: none;
  transition: none;
}

.c9.selected,
.c9[aria-selected=true] {
  background-color: #e4e9ee;
  box-shadow: inset 0 1px 0 rgba(208,215,222,0.2);
}

.c9:disabled,
.c9.disabled,
.c9[aria-disabled=true] {
  color: #8c959f;
  background-color: #f6f8fa;
  cursor: default;
}

.c9:focus,
.c9.focus {
  outline: none;
  box-shadow: 0 0 0 3px rgba(9,105,218,0.15);
}

.c9.danger:hover,
.c9.danger.hover,
[open] > .c9.danger {
  color: #ffffff;
  background-color: #a40e26;
  border-color: rgba(208,215,222,0.5);
  box-shadow: 0 1px 0 rgba(27,31,36,0.1), inset 0 1px 0 rgba(255,255,255,0.03);
}

.c9.danger:active,
.c9.danger.selected,
.c9.danger[aria-selected=true] {
  color: #ffffff;
  background-color: #c2202b;
  border-color: rgba(208,215,222,0.5);
  box-shadow: inset 0 1px 0 rgba(76,0,20,0.2);
}

.c9.danger:disabled,
.c9.danger.disabled,
.c9.danger[aria-disabled=true] {
  color: rgba(207,34,46,0.5);
  background-color: #f6f8fa;
  border-color: rgba(208,215,222,0.5);
  box-shadow: none;
}

.c9.danger:focus,
.c9.danger.focus {
  border-color: rgba(208,215,222,0.5);
  box-shadow: 0 0 0 3px rgba(164,14,38,0.4);
}

<div
  class="c0 selector-hello"
>
  <div
    class="c1"
  >
    <h2
      class="h2"
    >
      Hello
    </h2>
  </div>
  <div
    class="c2"
  >
    <div
      class="c3"
    >
      <div
        class="c4"
      >
        <div
          class="c5"
        >
          <div
            class="c6"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="c7"
  >
    <div
      class="c8"
    />
    <div
      class="c8"
    >
      <button
        class="c9 danger"
      >
        Close
      </button>
    </div>
  </div>
</div>
`;
